source/Irrlicht/COpenGLFunctionTable.cpp

```diff
@@ -0,0 +1,1500 @@
+			pGlGetIntegeri_v = (PFNGLGETINTEGERI_VPROC)IRR_OGL_SDL_LOAD_FUNCTIONPTR("glGetIntegeri_v");
+			pGlGetStringi = (PFNGLGETSTRINGIPROC)IRR_OGL_SDL_LOAD_FUNCTIONPTR("glGetStringi");
+
+			//should contain space-separated OpenGL extension names
+			constexpr const char* OPENGL_EXTS_ENVVAR_NAME = "_IRR_OPENGL_EXTENSIONS_LIST";//move this to some top-level header?
+
+			const char* envvar = std::getenv(OPENGL_EXTS_ENVVAR_NAME);
+			if (!envvar)
+			{
+				GLint extensionCount;
+				glGetIntegerv(GL_NUM_EXTENSIONS, &extensionCount);
+				for (GLint i = 0; i < extensionCount; ++i)
+				{
+					const char* extensionName = reinterpret_cast<const char*>(pGlGetStringi(GL_EXTENSIONS, i));
+
+					for (uint32_t j = 0; j < IRR_OpenGL_Feature_Count; ++j)
+					{
+						if (!strcmp(OpenGLFeatureStrings[j], extensionName))
+						{
+							FeatureAvailable[j] = true;
+							break;
+						}
+					}
+				}
+			}
+			else
+			{
+				std::stringstream ss{ std::string(envvar) };
+				std::string extname;
+				extname.reserve(100);
+				while (std::getline(ss, extname))
+				{
+					for (uint32_t j = 0; j < IRR_OpenGL_Feature_Count; ++j)
+					{
+						if (extname == OpenGLFeatureStrings[j])
+						{
+							FeatureAvailable[j] = true;
+							break;
+						}
+					}
+				}
+			}
+
+			float ogl_ver;
+			sscanf(reinterpret_cast<const char*>(glGetString(GL_VERSION)), "%f", &ogl_ver);
+			Version = static_cast<uint16_t>(core::round(ogl_ver * 100.0f));
+
+			GLint num = 0;
+			glGetIntegerv(GL_MAX_ELEMENTS_INDICES, &num);
+			MaxIndices = num;
+			glGetIntegerv(GL_MAX_ELEMENTS_VERTICES, &num);
+			MaxVertices = num;
+			glGetIntegerv(GL_MAX_VERTEX_STREAMS, &num);
+			MaxVertexStreams = num;
+			glGetIntegerv(GL_MAX_TRANSFORM_FEEDBACK_INTERLEAVED_COMPONENTS, &num);
+			MaxXFormFeedbackComponents = num;
+			glGetIntegerv(GL_MAX_SERVER_WAIT_TIMEOUT, &num);
+			MaxGPUWaitTimeout = reinterpret_cast<const uint32_t&>(num);
+
+			if (FeatureAvailable[IRR_NV_shader_thread_group])
+			{
+				glGetIntegerv(GL_WARP_SIZE_NV, &num);
+				InvocationSubGroupSize[0] = InvocationSubGroupSize[1] = reinterpret_cast<const uint32_t&>(num);
+			}
+			else if (IsIntelGPU)
+			{
+				InvocationSubGroupSize[0] = 4;
+				InvocationSubGroupSize[1] = 32;
+			}
+
+			if (FeatureAvailable[IRR_ARB_spirv_extensions])
+			{
+				glGetIntegerv(GL_NUM_SPIR_V_EXTENSIONS, reinterpret_cast<GLint*>(&SPIR_VextensionsCount));
+				if (SPIR_VextensionsCount)
+					SPIR_Vextensions = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<const GLubyte*> >(SPIR_VextensionsCount);
+				for (GLuint i = 0u; i < SPIR_VextensionsCount; ++i)
+					(*SPIR_Vextensions)[i] = pGlGetStringi(GL_SPIR_V_EXTENSIONS, i);
+			}
+			else
+			{
+				SPIR_VextensionsCount = 0u;
+			}
+
+			/**
+			pGl = () IRR_OGL_SDL_LOAD_FUNCTIONPTR("gl");
+			**/
+			pGlProvokingVertex = (PFNGLPROVOKINGVERTEXPROC)IRR_OGL_SDL_LOAD_FUNCTIONPTR("glProvokingVertex");
```

Comment:
My DLL Loader and DLL table C-preprocessor macros and classes that I've shown you, will do this automatically for you!

Reply:
So, the cast to whatever is in the parentheses is no longer required?
Im not sure if the macro generates any function within the class to initialize those pointers, if so, can this be done like this?
```
using LibLoader = system::DefaultFuncPtrLoader;
IRR_SYSTEM_DECLARE_DYNAMIC_FUNCTION_CALLER_CLASS(OpenGLMultiTexturing, LibLoader,
			glActiveTexture,
			//...
			)
OpenGLMultiTexturing multitexturing;
```

and in some function where i initialize the pointers i would use the `SDL_GL_GetProcAddress` in this fashion
```
multitexturing.pglActiveTexture = SDL_GL_GetProcAddress("glActiveTexture");
```

Reply:
> So, the cast to whatever is in the parentheses is no longer required?

It will do all the function pointer loading in the constructor "automagically".

> Im not sure if the macro generates any function within the class to initialize those pointers, if so, can this be done like this?

You need to change this out `using LibLoader = system::DefaultFuncPtrLoader;` for something else that you derive from the same base, because the default uses Windows/Unix dll system symbol retrieval functions, and for OpenGL special functions need to be used (I recommend the `SDL_GL_GetProcAddress`). 

Then everything is done automatically in constructors, then you don't need to do anything extra except for default constructing `multitexturing`.

I advise you use the special MSVC compiler flag `/P`
https://docs.microsoft.com/en-us/cpp/build/reference/p-preprocess-to-a-file?view=vs-2019
to see what these macros expand to (especially the already working OpenCL and CUDA implementations)